### PR TITLE
Jonas/refactor tables

### PR DESCRIPTION
**Pull request by jschneider-bensch**
This PR greatly simplifies table data structures and thereby table transformations.

It's still a draft for now since the WASM demo needs more help. It's updated to use the new table types but the `wasm` feature does not build at the moment, I believe because of the dependency on `libcrux`, although I tried enabling `libcrux`'es own `wasm` feature, to no avail.

### CODE REVIEW

**Comment by franziskuskiefer** [comment]
What's the state here?

**Comment by jschneider-bensch** [comment]
The demo can be built now since I pinned `libcrux` to an earlier commit  by using the `emscripten` toolchain, i.e.
```shell
$ cd ./hacspec-scrambledb/scrambledb
$ CC=emcc AR=emar wasm-pack build --target web --features wasm
```
builds the demo in `hacspec-scrambledb/scrambledb/pkg`.  I think the main part of the PR (the tables) are for sure ready for review and tidying up the demo build could be addressed in a follow-up.

File: hacspec-scrambledb/scrambledb/Cargo.toml [hunk 1]
```diff
@@ -21,7 +21,7 @@ rand = { version = "0.8.5", optional = true }
 getrandom = { version = "0.2.10", features = ["js"], optional = true }
 hex = { version = "0.4.3", optional = true }
 
-libcrux = { git = "https://github.com/cryspen/libcrux.git" }
+libcrux = { git = "https://github.com/cryspen/libcrux.git", rev = "8889c70b1faf26d131f14442f54a5938ab1deff6" }
```
Comment (franziskuskiefer) [inline hacspec-scrambledb/scrambledb/Cargo.toml:24]: Let's file an issue to update this to get back to main or a release.

File: hacspec-scrambledb/scrambledb/src/data_types.rs [hunk 2]
```diff
@@ -10,20 +10,26 @@ use oprf::coprf::coprf_online::{BlindInput, BlindOutput};
 
 /// A type for finalized pseudonyms, i.e. those which have been hardened for
 /// storage by applying a PRP.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(test, derive(Hash))]
```
Comment (franziskuskiefer) [inline hacspec-scrambledb/scrambledb/src/data_types.rs:14]: Do you want to keep this in only for tests or in general?

File: hacspec-scrambledb/scrambledb/src/join.rs [hunk 3]
```diff
@@ -123,49 +99,65 @@ pub fn join_conversion(
...
+                entry,
+                randomness,
+            )
+        })
+        .collect::<Result<Vec<BlindedPseudonymizedData>, Error>>()?;
+
+    converted_data.sort();
+
+    Ok(Table::new(table.identifier().into(), converted_data))
+    // for table in tables {
```
Comment (franziskuskiefer) [inline hacspec-scrambledb/scrambledb/src/join.rs:124]: What's with this commented code?

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Looks like this needs another pass to get cleaned up.

File: hacspec-scrambledb/scrambledb/src/data_types.rs (hunk 2 above)
Comment (jschneider-bensch) [reply hacspec-scrambledb/scrambledb/src/data_types.rs:14]: We only need it for building `HashSet`s during testing, so I left it like this.

File: hacspec-scrambledb/scrambledb/src/join.rs (hunk 3 above)
Comment (jschneider-bensch) [reply hacspec-scrambledb/scrambledb/src/join.rs:124]: Sorry, this is the old code. I'll make a another pass updating the docs on here as well.

**Comment by jschneider-bensch** [comment]
With the latest commit, I've removed the old comments, aligned the transformation names with the naming scheme from the spec and given more consistent doc comments.

**Review by franziskuskiefer — APPROVED** [review]
Thanks

**Pull request merged**